File: spot-service/src/main/kotlin/it/unibo/harmonikt/resources/SpotMarkers.kt
package it.unibo.harmonikt.resources

import io.ktor.resources.Resource
import kotlin.uuid.Uuid

/**
 * Represents the SpotMarkers resource in the Spot service.
 *
 * This resource is used to manage markers within the Spot service,
 * allowing for operations such as retrieving, creating, and managing
 * specific markers identified by their unique IDs.
 */
@Resource("/markers")
class SpotMarkers {
    /**
     * Represents a specific marker identified by its ID.
     *
     * This nested resource allows for operations on a specific marker
     * by its unique identifier.
     * @param parent The parent SpotMarkers resource.
     * @param id The unique identifier of the marker.
     */
    @Resource("{id}")
    class Id(val parent: SpotMarkers, val id: Uuid) {
        /**
         * Represents the position of a specific marker.
         *
         * This nested resource allows for operations related to the position
         * of a specific marker identified by its ID.
         * @param parent The parent Id resource of SpotMarkers.
         * @param description A description of the marker's position.
         */
        @Resource("position")
        class Position(val parent: Id, val description: String)
    }
}

File: spot-service-py/app/resources/spot_markers.py
"""
Spot Markers Resource Module

This module defines the resource structure for markers in the Spot service.
In FastAPI, we don't need to define resource classes like in Ktor, but we can
define the URL paths and route structures here for documentation purposes.
"""

from uuid import UUID

# Define the URL paths for the marker resources
MARKERS_PATH = "/markers"
MARKER_ID_PATH = "/markers/{marker_id}"
MARKER_POSITION_PATH = "/markers/{marker_id}/position"

# Define the route parameters for documentation
MARKER_ID_PARAM = {
    "marker_id": {"description": "The unique identifier of the marker", "type": UUID}
}


class SpotMarkers:
    """
    Represents the SpotMarkers resource in the Spot service.

    This resource is used to manage markers within the Spot service,
    allowing for operations such as retrieving, creating, and managing
    specific markers identified by their unique IDs.
    """

    class Id:
        """
        Represents a specific marker identified by its ID.

        This nested resource allows for operations on a specific marker
        by its unique identifier.
        """

        class Position:
            """
            Represents the position of a specific marker.

            This nested resource allows for operations related to the position
            of a specific marker identified by its ID.
            """

            pass
